lxd/instance/drivers/driver_qemu.go

```diff
@@ -1788,6 +1788,62 @@ func (d *qemu) deviceAttachBlockDevice(deviceName string, configCopy map[string]
 	return nil
 }
 
+func (d *qemu) deviceDetachBlockDevice(deviceName string, rawConfig deviceConfig.Device) error {
+	// Check if the agent is running.
+	monitor, err := qmp.Connect(d.monitorPath(), qemuSerialChardevName, d.getMonitorEventHandler())
+	if err != nil {
+		return err
+	}
+
+	escapedDeviceName := filesystem.PathNameEncode(deviceName)
+	deviceID := fmt.Sprintf("%s%s", qemuDeviceIDPrefix, escapedDeviceName)
+	blockDevName := fmt.Sprintf("%s%s", qemuBlockDevIDPrefix, escapedDeviceName)
+
+	err = monitor.RemoveDevice(deviceID)
+	if err != nil {
+		return err
+	}
+
+	waitDuration := time.Duration(time.Second * time.Duration(10))
+	waitUntil := time.Now().Add(waitDuration)
+	for {
+		err = monitor.RemoveBlockDevice(blockDevName)
+		if err == nil {
+			break
+		}
+
+		if err == qmp.ErrInUse {
+			time.Sleep(time.Second * time.Duration(2))
+			continue
+		}
+
+		if time.Now().After(waitUntil) {
+			return fmt.Errorf("Failed to detach block device after %v: %w", waitDuration, err)
+		}
+	}
+
+	if rawConfig != nil {
+		var newDevPaths []string
+
+		for _, path := range d.devPaths {
+			if path == rawConfig["source"] {
```

Comment:
This logic has the same problem as the original proposal of the `DevPath()` function, i.e it is assuming that for all disk types the `source` property available at non-start time is the same source path that was returned by the device's start-time RunConfig.

Now that we are recording into the json file the actual source returned from the device's start-time RunConfig we need to remove that value from the AppArmor profile, which is not necessarily the one in `source`.

Instead I suggest expanding the json file struct to look a bit like `NICConfig` struct, i.e containing `DeviceName` and then `SourcePath`. This way you can match on `DeviceName` and not include the associated `SourcePath` in the apparmor profile.

Reply:
Instead of introducing a new struct, I'd go with a `map[string]string`, the key being the device name and the value the source path.

Reply:
I would suggest using a new struct type as it allows for easier expandibility in the future should we need to add more fields. Also it has the added benefit that we can then easily see where that type is used, as compared to map[string]string which is used all over the place.